### DISCUSSION TITLE
[Language idea] Personify the compiler

### DISCUSSION THREAD

**Discussion by ConsciousCode** [hedyorg/hedy#3450] [Large Projects & Plans]
Specifically in the error messages, especially for earlier levels, I could see personification of the compiler as being an effective anchor for "why" certain things aren't allowed. 

For instance, "The code you entered is not valid Hedy code. There is a mistake on line 3, at position 18. You typed a comma, but that is not allowed." is a very technical and impersonal description. Much more descriptive than normal compiler messages, but my first thought when reading that is that many kids will wonder "why isn't it allowed?" 

On one hand that can be a gentle introduction to how languages have hundreds of grammar rules which you may not know when you first start, but on the other hand imagine an alternative - personify the compiler as some kind of silly robot that can't understand conversational language very well. Then, eg it isn't that the comma isn't "allowed", but the robot is confused by the comma and doesn't know what it means. 

This could serve as a gentle introduction to the idea that the compiler only knows a very restrictive grammar and it isn't enough to be "close enough", and it acts as a kind of empathic game to write the code in a way the silly robot understands.

**Comment by Felienne** [comment]
Just to clarify, this project would mean a large overhaul of the phrasing of error messages, so is not technically very complex but requires deep thinking about who the error message is written by. We have a bit of a mixed strategy now, we use all of these:

**Sometimes we ask a question: "Did you mean print?"**
<img width="698" alt="image" src="https://github.com/hedyorg/hedy/assets/1003685/709f9d65-6268-468e-b4fd-ae65acd0d4a0">

**Sometimes we give a suggestion "Add a print statement...."**
<img width="690" alt="image" src="https://github.com/hedyorg/hedy/assets/1003685/8487bd8b-762a-4f6a-8f47-0b5f942b5f1b">

**Sometimes, in case of a parse error, as described above, we just name the illegal character**
<img width="694" alt="image" src="https://github.com/hedyorg/hedy/assets/1003685/d5ab2a67-7caa-45a7-9c27-3c859e105533">

This project would entail firstly do think about what a good phrasing would be (thinking, reading papers, interview teachers?) and then make them all similar to what we want.

**Comment by jpelay** [comment]
> This could serve as a gentle introduction to the idea that the compiler only knows a very restrictive grammar and it isn't enough to be "close enough", and it acts as a kind of empathic game to write the code in a way the silly robot understands.

I quite like this idea, perhaps instead of the current symbol we use in alerts and error messages, we can add the icon of a robot?

![image](https://github.com/hedyorg/hedy/assets/45865185/1febcd22-566d-46b8-9ebb-cc7cc6766ecd)
